### DISCUSSION TITLE
Need to create page in external module

### DISCUSSION THREAD

**Discussion by Gotiap** [oqtane/oqtane.framework#1363] [Q&A]
Hello,

I would like to add page in pages folder as we already have in **Oqtane.Server/Pages/Login.cshtml**, The same way I would like to add it into ExternalProject.Server/Pages/XYZ.cshtml.

Basically I want to use OnPostAsync method from external modules for my created page. But it does not allow any such routing like  "/{alias}/pages/login". Even I do not want to customize routing or anything I just want to use OnPostAsync method of any page which I can create into external module.

Thank you for your help.

**Comment by sbwalker** [comment]
It would be helpful to understand what you are trying to accomplish by using razor pages on the server. The only reason why the login and logout pages exist in the framework are to facilitate the transfer of the auth cookie from the server to the client when running on Blazor Server.

**Reply by Gotiap** [reply]
As I want to customize login module to use OIDC. So I need to use form POST method.

**Reply by sbwalker** [reply]
Oqtane is multi-tenant, so it is essential that tenant context is transferred from the client to the server when making HTTP calls. The way this is accomplished is by including the AliasId as a prefix on your route. This is why the Login and Logout pages use a route in the form of "/{alias}/pages/login". Note that in 2.1.0 (not yet released) this approach has been changed and HTTP calls from the client to the server are now based on the human friendly alias value rather than AliasId ( ie. www.domain.com or www.domain.com/department or www.domain.com/department/group, etc... ).

**Reply by Gotiap** [reply]
Yes, That I understood.
If I create any pages into oqtane this works fine, But for that actually I need to create page into oqtane.server project. So it depends on oqtane module not completely external module.

So my main aim is that I can use alias as well for call that page and call get and post method on the page which I created on external modules instead of in oqtane.server/pages folder.

Is it something possible?

**Reply by sbwalker** [reply]
This SO post provides a hint: 

https://stackoverflow.com/questions/53954850/how-to-dynamically-load-pages-from-plugins-in-razor-pages-application 

Note that Oqtane already dynamically loads ApplicationParts in module assemblies:

                // check if assembly contains MVC Controllers
                if (assembly.GetTypes().Any(t => t.IsSubclassOf(typeof(Controller))))
                {
                    var partFactory = ApplicationPartFactory.GetApplicationPartFactory(assembly);
                    foreach (var part in partFactory.GetApplicationParts(assembly))
                    {
                        mvcBuilder.PartManager.ApplicationParts.Add(part);
                    }
                }

So my question is if you are certain that your .Views.dll file for your external module is being deployed to the Oqtane /bin folder. The standard configuration in the external module /Package folder does not handle Views.dll file - so you would need to edit the debug.cmd and/or *.nuspec

**Reply by Gotiap** [reply]
Okay, Thank you @sbwalker ,
Will check that and let you know on the same.

**Reply by sbwalker** [reply]
@Gotiap were you able to confirm that if you includes the *.Views.dll file then Oqtane will load it?

**Reply by Gotiap** [reply]
@sbwalker 

Thank you for your reply. I tried to add that but basically, views.dll does not contain any controller, So I need to update the code a bit and added IsSubclassOf(typeof(Microsoft.AspNetCore.Mvc.RazorPages.Page)).
After that, it picked up.

```
if (assembly.GetTypes().Any(t => t.IsSubclassOf(typeof(Controller)) || t.IsSubclassOf(typeof(Page))))
                {
                    var partFactory = ApplicationPartFactory.GetApplicationPartFactory(assembly);
                    foreach (var part in partFactory.GetApplicationParts(assembly))
                    {
                        mvcBuilder.PartManager.ApplicationParts.Add(part);
                    }
                }

```                
Thank you again.

**Reply by sbwalker** [reply]
@Gotiap this does seem to be a useful change to make to the framework so that it handles razor pages - do you want to submit a PR?

**Reply by Gotiap** [reply]
Hello @sbwalker 

Yup, Sure I'll raise PR.

Thanks.

**Reply by vzdesic** [reply]
> As I want to customize login module to use OIDC. So I need to use form POST method.

Hello @Gotiap! Beside this feature to support pages in external module, can you please say few more words about your plan to implement OIDC. I also want to implement OIDC against IdentityServer4. How do you plan to inject OIDC service? Do you plan to send direct POST request to authorize endpoint? Do you have concrete scenario?

Thanks a lot in advance!
